pkg/frontend/clustermanager_putorpatch.go

```diff
@@ -102,52 +120,369 @@ func (f *frontend) _putOrPatchClusterManagerConfiguration(ctx context.Context, l
 		if ocmdoc.Deleting {
 			return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeRequestNotAllowed, "", "Request is not allowed on a resource marked for deletion.")
 		}
-		if ocmdoc.ClusterManagerConfiguration != nil {
-			ocmdoc.ClusterManagerConfiguration.Properties.Resources = []byte(resources)
+		ocmdoc.SyncSet.Properties.Resources = string(resources)
+	}
+
+	ocmdoc.CorrelationData = correlationData
+	f.systemDataSyncSetEnricher(ocmdoc, systemData)
+
+	ocmdoc, err = f.dbClusterManagerConfiguration.Update(ctx, ocmdoc)
+	if err != nil {
+		return nil, err
+	}
+
+	ext := converter.ToExternal(ocmdoc.SyncSet)
+	b, err := json.MarshalIndent(ext, "", "  ")
+	return b, err
+}
+
+func (f *frontend) _putOrPatchMachinePool(ctx context.Context, log *logrus.Entry, r *http.Request, header *http.Header, converter api.MachinePoolConverter, staticValidator api.ClusterManagerStaticValidator) ([]byte, error) {
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	correlationData := r.Context().Value(middleware.ContextKeyCorrelationData).(*api.CorrelationData)
+	systemData, _ := r.Context().Value(middleware.ContextKeySystemData).(*api.SystemData) // don't panic
+	vars := mux.Vars(r)
+
+	_, err := f.validateSubscriptionState(ctx, r.URL.Path, api.SubscriptionStateRegistered)
+	if err != nil {
+		return nil, err
+	}
+
+	originalPath := r.Context().Value(middleware.ContextKeyOriginalPath).(string)
+	armResource, err := arm.ParseArmResourceId(originalPath)
+	if err != nil {
+		return nil, err
+	}
+
+	ocp, err := f.dbOpenShiftClusters.Get(ctx, armResource.ParentResource())
+	if err != nil && !cosmosdb.IsErrorStatusCode(err, http.StatusNotFound) {
+		return nil, err
+	}
+
+	if ocp == nil || cosmosdb.IsErrorStatusCode(err, http.StatusNotFound) {
+		return nil, api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	}
+
+	ocmdoc, _ := f.dbClusterManagerConfiguration.Get(ctx, r.URL.Path)
+	if err != nil && !cosmosdb.IsErrorStatusCode(err, http.StatusNotFound) {
+		return nil, err
+	}
+
+	var resources string
+	err = json.Unmarshal(body, &resources)
+	if err != nil {
+		return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized: %q.", err)
+	}
+
+	err = staticValidator.Static(resources, vars)
+	if err != nil {
+		return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The 'Kind' in the request payload does not match 'Kind' in the request path: %q.", err)
+	}
+
+	isCreate := ocmdoc == nil
+	uuid := f.dbClusterManagerConfiguration.NewUUID()
```

Comment:
I believe this chunk is duplicated 4 times in this file. Simply refactoring this chunk out leaves much to be desired as there are way too many return values. I played around a little and was able to extract lines 145-163 into a function returning an error and the 'originalPath'. 

```go
func validateOriginalPath(ctx context.Context, r *http.Request, vars map[string]string) (string, error) {
	_, err := f.validateSubscriptionState(ctx, r.URL.Path, api.SubscriptionStateRegistered)
	if err != nil {
		return "", err
	}

	originalPath := r.Context().Value(middleware.ContextKeyOriginalPath).(string)
	armResource, err := arm.ParseArmResourceId(originalPath)
	if err != nil {
		return "", err
	}

	ocp, err := f.dbOpenShiftClusters.Get(ctx, armResource.ParentResource())
	if err != nil && !cosmosdb.IsErrorStatusCode(err, http.StatusNotFound) {
		return "", err
	}

	if ocp == nil || cosmosdb.IsErrorStatusCode(err, http.StatusNotFound) {
		return "", api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
	}
	return originalPath, err
}
```

Reply:
Thanks, agreed that this is a good chunk to DRY. I've worked in your changes.